lib/node_searcher.rb

```diff
@@ -0,0 +1,61 @@
+require_relative "./stack.rb"
+
+class NodeSearcher
+
+  attr_accessor :tree
+
+  def initialize
+    @tree = nil
+    @stack = Stack.new
+  end
+
+  def search_by(query_type, query)
+    @stack.push(@tree.root)
+    result_array = []
+    until @stack.empty?
+      current_node = @stack.pop
+      current_node.children.each do |child|
+        @stack.push(child)
+        #binding.pry
+        if (child.data[query_type]) && (child.data[query_type].include? query)
+          result_array << child
+        end
+      end
+    end
+    result_array   
+  end
+
+  def search_children(node, query_type, query)
+    @stack.push(node)
+    result_array = []
+    until @stack.empty?
+      current_node = @stack.pop
+      current_node.children.each do |child|
+        @stack.push(child)
+        if (child.data[query_type]) && (child.data[query_type].include? query) &&
+          child.depth > node.depth
```

Comment:
This is the only line of code that's different from the function above, "search_by". Plus, it's very similar to line 53 below. It would be super-slick to combine all 3 into one re-usable function.